### ISSUE TITLE
teach tmt run discover to use a condition besides a filter

### ISSUE THREAD

**Issue by sopos** [teemtee/tmt#373]
as `filter` is sometimes not enough it would be good to be able to use --condition the same as for `tmt tests ls`

**Comment by sopos** [comment]
it would be actually useful to have this available also on command line

**Comment by sopos** [comment]
probably all the possibilities from the plan should be possible to set form command line..

**Comment by sopos** [comment]
I'm a bit confused. looks like I was looking at the wrong place. It is not a `discover` what I'm looking for but `run tests`. But `discover -h tmt` has the `--filter` ability as well.
@psss can you clarify the difference and the basic use cases? Can I put `tests:` as kind of _step_ into the plan as an equivalent to `tests` selector on command line?

**Comment by psss** [comment]
The `discover` step is used as the primary source of the information where tests should be explored and how they should be selected (filtered). The `tmt run test` subcommand allows to apply additional filters from the command line. It is used primarily for selecting tests for debugging without having to modify plans. It works also across `discover` step implementations, for example:

    tmt run discover -v plan --name helps test --name smoke

Is it a bit more clear now?

**Comment by sopos** [comment]
I see, so my request to be able to use _condition_ as well as _filter_ is still valid. Actually, _condition_ allow us to do more advanced filtering than _filter_ itself.

**Comment by psss** [comment]
Yes, but the important question is if we really want to support `condition` in L2 metadata as well. Because the feature runs python's `eval` so it can be quite dangerous. Basically arbitrary code execution. Which might make sense if you explicitly request from the command line but if this could be included in config files, then it would be no more safe to run `tmt run` on your laptop... We need to properly discuss this.

**Comment by hegerj** [comment]
That is not ideal but then `--filter` functionality must be extended.

**Comment by i386x** [comment]
@psss Could `eval`-based constructs be replaced by Jinja2 templates? Jinja2 expressions are much more safer than raw Python code and they have almost same syntax and semantics. Additionally, if some feature is missing, it can be implemented and passed to Jinja2 context so it can be then used in Jinja2 expressions. Jinja2 expressions are for example used by Ansible in Ansible playbooks where they play important role in conditionals and as a value generators.

**Comment by i386x** [comment]
Just to complete my previous comment and for inspiration, Ansible uses a kind of "safe" eval implementation: https://github.com/ansible/ansible/blob/devel/lib/ansible/template/safe_eval.py#L31

**Comment by jscotka** [comment]
I also miss `--condition` option also in Metadata(plans) as well :-) not from perspective that I've created that idea for FMF but that filter is very limited, and I have troubles to write some items like: `"some" in tag and not "another" in tag` or `re.search("some.*test", description)`

Jinja is propbably better than nothing, but not sure if I'll be able to do some advanced usecases, e.g. go deep into structure of metadata like `data['enviromnent']['PYTHON']="3.7"` or `"some" in data['key']['key2']`

**Comment by jscotka** [comment]
theoretically I can imagine some restrictions like described here:
https://realpython.com/python-eval-function/#minimizing-the-security-issues-of-eval
 and pass just allowed locals(), globals()  via these suggestions.

or e.g. I can imagine also something like call this eval inside something like `unshare -n -p -m` not sure if there is some equivalent inside python without privileged users, like block net/file/proc ops with some mocking:  `import socket; socket.socket = None`

**Comment by psss** [comment]
This one has been stalled for some time. Is the use case still relevant? Do we find enough benefit in addressing this? The security aspect/question could be mitigated by only allowing the extra filtering under `--feeling-safe` mode.

**Comment by psss** [comment]
@sopos, would you still welcome this functionality?
@mkluson has some use cases for filtering plans not only by name. Is this something which could be possibly beneficial for your use case?

**Comment by psss** [comment]
After a quick discussion with @sopos and @mkluson there seems to be no more interest in this. Closing. Anyone feel free to reopen if you still think it would be useful.
